Sprint 3

Sapr0nov opened the pull request:
добавлено:
Router, WebSocket, API
тестирование Mocha, Chai
безопасность helmet и express-rate-limit

Sapr0nov committed:
обновление пакетов, вынос сборки fetch перенастройка роутов

Sapr0nov committed:
форма регистрации обновлена валидация, добавлено api

Sapr0nov committed:
подключен EventBus завершена работа с пользвателем по API

Sapr0nov committed:
загрузка аватарки

Sapr0nov committed:
добавлены функции для обновления пароля

Sapr0nov committed:
конетекстное меню, работа с сокетами добавление удаление пользователей в чат

Sapr0nov committed:
Подключен WebSocket

Sapr0nov committed:
connect Tests Mocha

Sapr0nov committed:
Переписан под роутер отдельным модулем для тестов...

Sapr0nov committed:
дописаны тесты с поднятием сервера и настроена безопасность

Sapr0nov committed:
fixed routing зафиксированы версии пакетов

Sapr0nov committed:
routing netlify

agadoev commented on package.json:
```diff
@@ -1,35 +1,53 @@
+    "dev": "parcel ./src/index.html --no-source-maps",
+    "test": "./node_modules/.bin/mocha --require ts-node/register",
+    "build": "parcel build ./src/index.html --no-source-maps",
+    "buildAll": "parcel build ./src/index.html --no-source-maps && parcel build ./src/modules/Fetch.ts --no-source-maps --no-scope-hoist --no-optimize --no-content-hash",
```
Думаю, что тут стоит добавить скрипт для запуска eslint'a и stylelint'a. А то зависимости есть, но не используются

agadoev commented on src/modules/form.css:
```diff
@@ -64,6 +64,8 @@
 }
```
в одной папке не должны находиться и .ts и .css файлы. Как вариант, можешь создать папку assets или public и хранить там стили

agadoev commented on src/modules/route.spec.ts:
```diff
@@ -0,0 +1,20 @@
+import { Block } from "./Block";
```
Есть отдельная папка tests, но почему-то тесты лежат в исходниках. Логичнее было бы поместить их как раз в папку tests

Sapr0nov replied:
Согласен это было бы логично, но оставлю как есть т.к. это было в официальном требовнии п.5 "Например, тесты для роутера должны лежать в той же папке, что и файл с кодом роутера. Используйте Mocha и Chai. С Jest вы будете работать в следующем модуле."

agadoev commented on src/modules/Validate.ts:
```diff
@@ -4,88 +4,109 @@ interface FormData {
-    isPhone(str:string) {
-        const regEx = /^\+?[0-9]*/gm;
-        if (!regEx.test(str)) { throw "может начинаться с плюса, состоит только из цифр"; };
+    isPhone(str:string):string|boolean {
+        const regEx = /^\+?[0-9]*$/gm;
+        if (!regEx.test(str)) { 
+            return "может начинаться с плюса, состоит только из цифр"; 
+        }
+        return true;
     }
```
Валидатор не должен возвращать строку, он должен возвращать только boolean. Существует принцип единственной ответственности (SRP), который гласит, что метод (класс, функция, модуль и тд) должен иметь одну и только одну причину для изменения. Сейчас я могу изменить этот метод по двум причинам: 
1) Если хочу изменить текст ошибки
2) Если хочу изменить правило валидации

То, что отображается на экране никак не должно быть связано с валидацией и должно находидиться на уровне отображения, т.е. в компоненте, который рендерит шаблон

Sapr0nov replied:
Поправил, хотя не совсем согласен, здесь идет описание того, что проверяется, а не текст ошибки. Которые потом формируется опираясь на это описание. И  если мы используем этот код в нескольких местах, нам несколько раз приходиться дублировать это описание, что противоречит принципу DRY.  В дальнейшем я вижу развите этого кода как подстановка шаблонной строки i18 и уже в файлах i18 мы меняем описание и что проверяется и самой ошибки... Но в любом случае поправил и убрал. Ксати в предыдущей версии здесь был возврат ошибки через try catch, убрал по твоей просьбы, но сейчас могу еще сказать, что тоже на мой взгляд тот подход (подсмотренный где-то в учебных материалах) больше ложится в концепцию TypeScript.

agadoev commented on src/modules/Fetch.ts:
```diff
@@ -1,19 +1,26 @@
-    
+export default class Fetch {
+    ctxFetch = this;
     METHODS:MethodsCollection = {
         GET: 'GET',
 		POST: 'POST',
```
```suggestion
    Method: TMethod
```

agadoev commented on src/modules/Fetch.ts:
```diff
@@ -1,19 +1,26 @@
-    POST:Methods,
-    PUT:Methods,
-    DELETE:Methods
+export type Methods = 'GET' | 'POST' | 'PUT' | 'DELETE';
+
+export type MethodsCollection = {
+    GET: Methods,
+    POST: Methods,
+    PUT: Methods,
+    DELETE: Methods
```
Я думаю, что можно спокойно обойтись без MethodsCollection. Использование union type на первой строке - отличный вариант. Сейчас MethodsCollection выглядит лишним кодом.

agadoev commented on src/modules/Fetch.ts:
```diff
@@ -50,24 +57,26 @@ export class Fetch {
             const xhr = new XMLHttpRequest();
-            const isGet = method === this.METHODS.GET;
-
+            const isGet = (method === ctxFetch.METHODS.GET);
```
```suggestion
            const isGet = (method === 'GET');
```

agadoev commented on src/modules/Fetch.ts:
```diff
@@ -1,19 +1,26 @@
-export class Fetch {
-    
+export default class Fetch {
+    ctxFetch = this;
```
Кажется, что это поле можно убрать и обойтись без него. Чем меньше кода - тем меньше ошибок можно допустить

agadoev commented on src/app.ts:
```diff
@@ -1,76 +1,358 @@
+                };
+
+                const formData = new FormData(form as HTMLFormElement);
+                const data = api.signUp(
+                    formData.get('first_name') as string,
+                    formData.get('second_name') as string,
+                    formData.get('login') as string,
+                    formData.get('email') as string,
+                    formData.get('phone') as string,
+                    formData.get('password') as string
+                );
+                data.then( (req:XMLHttpRequest) => {
+                    const JSONreqest:UserFields = JSON.parse(req.response);
+                    if (JSONreqest.id) {
+                        router.go('/messenger');
+                    }
+                });
```
1) api.signUp - асинхронный метод и должен вызываться через await
2) Что, если сервер вернет ошибку? Например 500. Надо такие случаи перехватывать и выводить информацию об ошибке пользователю, хотя бы через alert()

Sapr0nov replied:
Он и вызывается в самом методе с ( await fetch )  и возвращает промис - сам же выделил код ниже `data.then...` где он и обрабатывает данные - пришли идем на страницу не чата, не пришли - вводи ещё )
При попытке исправить и поставить await выдает ошибку - 'await' expressions are only allowed within async functions and at the top levels of modules.

agadoev replied:
Да, правильно. await можно использовать только в асинхронных функциях, для этого в addEventListener нужно передать асинхронную функцию с ключевым словом async
Примерно так будет выглядеть
<img width="535" alt="image" src="https://user-images.githubusercontent.com/37894659/187906164-7a613760-5c7a-4f6a-a5d4-47e0bc59c479.png">

Хороший UI/UX обладает отзывчивостью, т.е. на каждое действие пользователь должен получить результат о его действии. Если произошло ошибка или например пользовательский ввод неправильный, то надо как-то уведомить пользователя об этом. Это больше про UI/UX история, но фронтендеру тоже значит полезно

Sapr0nov replied:
Ты абсолютно прав, в голове даже всплывает что, где-то, когда-то и я про это читал, так что ничего не могу возразить, добавлю в TODO 
p.s. Спасибо за пример!

agadoev commented on src/app.ts:
```diff
@@ -1,76 +1,358 @@
+
+                const formData = new FormData(form as HTMLFormElement);
+                api.signIn(
+                    formData.get('login') as string,
+                    formData.get('password') as string
```
Часто явное приведение типа - знак того, что где-то что-то пошло не так. Скорее всего этот код должен находиться не здесь, а в компоненте, к которому эта форма относится

agadoev commented on src/modules/validate.spec.ts:
```diff
@@ -0,0 +1,103 @@
+    });
+})
+
+const error5 = "Должна быть хотя бы одна заглавная буква";
+const desc5 = "ValidateForm: Пароль - Должна быть хотя бы одна заглавная буква";
```
1) Если вдруг бизнес захочет изменить текст об ошибке или например вставить туда ссылку, придется бегать по всем местам в коде, где хранится этот текст. Это не всегда удобно. Текст - это отображение и он должен находиться в компоненте, который рендерит шаблон.
2) Давай названия с цифрами - не очень хороший способ описать, что находится в переменной

agadoev reviewed (changes requested):
Привет
Основные моменты, над которыми стоит поработать
1) Структура проекта - сейчас почти весь код находится в папке modules. Ты можешь создать разные папки - core, utils, helpres, api, controllers, services, что захочешь. Чем сильнее ты дробишь свой проект, тем легче с ним будет потом работать
2) Сейчас очень много логики в app.ts. Ее точно нужно распутывать и помещать в компоненты/сервисы. Когда все в одном файле/функции, код получается слишком связанным и мы оказываемся в такой ситуации, что чтобы изменить какую-то одну вещь, нужно помнить, что мы можем что-то где-то сломать. Эта лишняя когнитивная нагрузка, которая делает работу над проектом сложнее

Посоветую 3 источника, которые смогут помочь (мне помогли)
1) Книга "Чистый код" Роберта Мартина
2) Книга "Чистая архитектура" Роберта Мартина
3) Доклад "Simple Made Easy" Ричи Хикки 

Пока что код не могу принять, давай попробуем его улучшить и посмотрим, как двигаться дальше

Bezumstvo commented:
Спасибо, за быстрый ответ, в условиях очередного сорваного дедлайна это бесцено. С доработками согласен в большинстве, но не понял какое основное условие для принятия работы я не выполнил?  опирался на эти пункты - https://practicum.yandex.ru/learn/middle-frontend/courses/3e08c3b3-ed6e-4483-a5c9-6cda18f0c726/sprints/69891/topics/bccb0fbe-eee9-4426-8004-9d2f4d64d732/lessons/e3b1630a-c6e6-4571-89e5-79d465e9f117/

agadoev commented:
Все задания выполнены, однако я не могу принять работу, пока у меня есть столько вопросов к коду

agadoev commented:
Сейчас писал код по работе и вспомнил принцип один хороший, который может и тебе пригодится
Когда пишешь код, сразу продумывай как его тестировать. Если его сложно протестировать, то скорее всего код спроектирован не очень хорошо и его стоит переписать. На эту тему можно почитать книгу "Экстремальное программирование: разработка через тестирование" Кента Бека

Bezumstvo commented:
> Сейчас писал код по работе и вспомнил принцип один хороший, который может и тебе пригодится Когда пишешь код, сразу продумывай как его тестировать. Если его сложно протестировать, то скорее всего код спроектирован не очень хорошо и его стоит переписать. На эту тему можно почитать книгу "Экстремальное программирование: разработка через тестирование" Кента Бека

Это я уже осознал)) когда пришлось переписывать половину после попыток написать тесты....

Sapr0nov committed:
исправление замечаний. Linter, Валидация, css вынесены и пр

Sapr0nov committed:
new file:   .eslintignore
	new file:   .eslintrc.json
	modified:   .mocharc.json
	modified:   .prettierrc.json
	new file:   .stylelintrc.json
	modified:   README.md
	modified:   package-lock.json
	modified:   package.json
	modified:   src/components/aside_chat_preview/aside_chat_preview.css
	modified:   src/components/aside_chat_preview/aside_chat_preview.hbs
	modified:   src/components/aside_chat_preview/aside_chat_preview.ts
	modified:   src/components/chat/chat.css
	modified:   src/components/chat/chat.hbs
	modified:   src/components/chat/chat.ts
	modified:   src/components/glitch_effect/glitch_effect.css
	modified:   src/components/settings/settings.css
	modified:   src/components/settings/settings.ts
	modified:   src/components/side_bar/side_bar.css
	modified:   src/components/side_bar/side_bar.hbs
	modified:   src/components/side_bar/side_bar.ts
	modified:   src/handlebars.d.ts
	modified:   src/index.html
	modified:   src/modules/ApiAction.ts
	modified:   src/modules/Block.ts
	modified:   src/modules/Chat.ts
	modified:   src/modules/Cookie.ts
	modified:   src/modules/EventBus.ts
	modified:   src/modules/Fetch.ts
	modified:   src/modules/Route.ts
	modified:   src/modules/Router.ts
	modified:   src/modules/User.ts
	modified:   src/modules/Validate.ts
	deleted:    src/modules/color.css
	deleted:    src/modules/form.css
	modified:   src/modules/route.spec.ts
	modified:   src/modules/validate.spec.ts
	modified:   src/pages/404/404.css
	modified:   src/pages/404/404.hbs
	modified:   src/pages/404/404.ts
	modified:   src/pages/500/500.css
	modified:   src/pages/500/500.hbs
	modified:   src/pages/500/500.ts
	modified:   src/pages/form_auth/form_auth.hbs
	modified:   src/pages/form_auth/form_auth.ts
	modified:   src/pages/form_registration/form_registration.hbs
	modified:   src/pages/form_registration/form_registration.ts
	modified:   src/pages/index/index.css
	modified:   src/pages/index/index.hbs
	modified:   src/pages/index/index.ts
	modified:   src/srv.js
	modified:   src/srv.spec.ts
	new file:   static/color.css
	modified:   static/data.json
	new file:   static/form.css
	modified:   test/babel-register.js
	modified:   test/index.spec.ts
	modified:   tsconfig.json

agadoev commented:
Привет)
Спасибо за правки, думаю, что код стал чище. Иногда бывает полезно вернуться к коду и посмотреть на него свежим взглядом. Чаще всего решение в лоб не всегда правильное и я, например, часто перепроверяю с утра написанный мною вчера код, потому что я могу найти места, которые стоит переделать. Думаю это полезная привычка, поэтому правки - это нормально)
В любом случая я работу принимаю, удачи в четвертом спринте :)

agadoev reviewed (approved):
Привет)
Спасибо за правки, думаю, что код стал чище. Иногда бывает полезно вернуться к коду и посмотреть на него свежим взглядом. Чаще всего решение в лоб не всегда правильное и я, например, часто перепроверяю с утра написанный мною вчера код, потому что я могу найти места, которые стоит переделать. Думаю это полезная привычка, поэтому правки - это нормально)
В любом случая я работу принимаю, удачи в четвертом спринте :)
